app/code/community/Bolt/Boltpay/Controller/Traits/ApiControllerTrait.php

```diff
@@ -0,0 +1,109 @@
+     * @var bool mandates that all request to this controller must be signed
+     */
+    protected $requestMustBeSigned = true;
+    /**
+     * For JSON, clears response body and header, and sets headers.
+     * After this, verifies request is from Bolt, if not, sends error message response
+     *
+     * @return Mage_Core_Controller_Front_Action
+     *
+     * @throws Exception Thrown if request cannot be verified as originating from Bolt
+     */
+    public function preDispatch()
+    {
+        @ob_start();
+        if ($this->willReturnJson) {
+            $this->getResponse()->clearAllHeaders()->clearBody();
+            $this->boltHelper()->setResponseContextHeaders();
+            $this->getResponse()
+                ->setHeader('Content-type', 'application/json', true);
+            $this->getLayout()->setDirectOutput(true);
+        }
+        if ( $this->requestMustBeSigned ) {
+            if (empty($this->payload)) { $this->payload = file_get_contents('php://input'); }
+            if (empty($this->signature)) { $this->signature = @$_SERVER['HTTP_X_BOLT_HMAC_SHA256']; }
+            $this->verifyBoltSignature($this->payload, $this->signature);
+        }
+        return parent::preDispatch();
+    }
+    /**
+     * Verifies that a request originated from and was signed by Bolt.  If not,
+     * an error response is sent to caller and the execution of the script is halted
+     * immediately
+     *
+     * @param string $payload       The body to be compared against a signature
+     * @param string $signature     The signature against the payload with the signing secret
+     *
+     * @throws Zend_Controller_Response_Exception if an invalid HTTP response code is set
+     */
+    protected function verifyBoltSignature($payload, $signature ) {
+        if (!$this->boltHelper()->verify_hook($payload, $signature)) {
+            $exception = new Bolt_Boltpay_OrderCreationException(
+                Bolt_Boltpay_OrderCreationException::E_BOLT_GENERAL_ERROR,
+                Bolt_Boltpay_OrderCreationException::E_BOLT_GENERAL_ERROR_TMPL_HMAC
+            );
+            $this->getResponse()
+                ->setHttpResponseCode($exception->getHttpCode())
+                ->setBody($exception->getJson())
+                ->setException($exception)
+                ->sendResponse();
+            $this->boltHelper()->notifyException($exception, array(), 'warning');
+            exit;
+        }
+    }
+    /**
+     * POST data in response to a request
+     *
+     * @param int   $httpCode       standard HTTP response code
+     * @param array $data           a PHP array to be encoded as JSON and sent as a response body to Bolt
+     * @param bool  $doJsonEncode   instructs whether to JSON encode the data, true by default
+     *
+     * @throws Zend_Controller_Response_Exception if the error code is not within the valid range
+     */
+    protected function sendResponse($httpCode, $data = array(), $doJsonEncode = true )
```

Comment:
I think it'll be simpler if we remove $doJsonEncode and always do json encoding

Reply:
`$doJsonEncode` is an option parameter and last in the list of parameters.  It can be ignored in all cases and conveniently produce an encoded JSON response.  Removing it adds no value.  Keeping it adds the value of removing the need for silly code in the case that json exists that is already encoded.  We don't want to be encouraging this sort of code.

`$this-sendResponse( 422, json_decode({"point":"This is silly"}) );`

Reply:
but you can write

> $this-sendResponse( 422, array("point" => "This is not silly") ); 

right? 

I don't think human should write raw JSON. Let's write object and let `json_encode` function handle JSON encoding. 
Within this PR, all instances you already use object or you can easily replace it with object

Reply:
ok. no problem.